fetcher.go

```diff
@@ -0,0 +1,173 @@
+package fetcher
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/johngibb/fetcher/lru"
+)
+
+// batch represents a single fetch.
+type batch struct {
+	ids     map[int64]struct{} // set of ids in batch
+	runOnce sync.Once          // ensure batch is only fetched once
+	done    chan struct{}      // signal that batch is done
+
+	results map[int64]interface{} // results from batch
+	err     error                 // error from batch
+}
+
+func newBatch() *batch {
+	return &batch{
+		done:    make(chan struct{}),
+		ids:     make(map[int64]struct{}),
+		results: make(map[int64]interface{}),
+	}
+}
+
+type cacheEntry struct {
+	val     interface{}
+	expires time.Time
+}
+
+// Fetcher caches and batches requests.
+type Fetcher struct {
+	wait      time.Duration                        // max time to wait for batch to fill up
+	batchSize int                                  // once batch exceeds this size, it's always run
+	fetch     func([]int64) ([]interface{}, error) // fetch function
+	cacheTTL  time.Duration                        // cache TTL; zero disables caching.
```

Comment:
I'd consider either treating cacheTTL==0 as no expiry or simplifying the code path for disabling the cache. Right now you have to test that cacheTTL==0 in more than one spot, and it even goes through the "batcher" code path. It _might_ make sense to change Fetcher.Get to this:

```go
func (f *Fetcher) Get(ctx context.Context, ids []int64) ([]interface{}, error) {
    if f.cacheTTL == 0 {
        return f.fetch(ids) // i.e. bypass the entire caching code path
    }
    // as before...
}
```

Reply:
This would bypass both caching _and_ batching though, right? I'd still want batching to occur when caching is disabled (basically falling back to the traditional [dataloader](github.com/facebook/dataloader) pattern).